python/ray/dag/tests/experimental/test_torch_tensor_dag.py

```diff
@@ -886,6 +910,949 @@ def test_torch_tensor_exceptions(ray_start_regular):
+    """
+    Test basic all-reduce.
+    """
+    if not USE_GPU:
+        pytest.skip("NCCL tests require GPUs")
+
+    assert (
+        sum(node["Resources"].get("GPU", 0) for node in ray.nodes()) > 1
+    ), "This test requires at least 2 GPUs"
+
+    actor_cls = TorchTensorWorker.options(num_cpus=0, num_gpus=1)
+
+    num_workers = 2
+    workers = [actor_cls.remote() for _ in range(num_workers)]
+
+    with InputNode() as inp:
+        computes = [
+            worker.compute_with_tuple_args.bind(inp, i)
+            for i, worker in enumerate(workers)
+        ]
+        collectives = collective.allreduce.bind(computes, ReduceOp.SUM)
+        recvs = [
+            worker.recv.bind(collective)
+            for worker, collective in zip(workers, collectives)
+        ]
+        dag = MultiOutputNode(recvs)
+
+    compiled_dag = dag.experimental_compile()
+
+    base_sum = (1 + num_workers) * num_workers / 2
+    for i in range(3):
+        i += 1
+        shape = (i * 10,)
+        dtype = torch.float16
+        ref = compiled_dag.execute(
+            [(shape, dtype, i + idx + 1) for idx in range(num_workers)]
+        )
+        result = ray.get(ref)
+        reduced_val = base_sum + num_workers * i
+        assert result == [(reduced_val, shape, dtype) for _ in workers]
+
+    compiled_dag.teardown()
+
+
+@pytest.mark.parametrize("ray_start_regular", [{"num_cpus": 4}], indirect=True)
+def test_torch_tensor_nccl_all_reduce_get_partial(ray_start_regular):
+    """
+    Test getting partial results from all-reduce does not hang.
+    """
+    if not USE_GPU:
+        pytest.skip("NCCL tests require GPUs")
+
+    assert (
+        sum(node["Resources"].get("GPU", 0) for node in ray.nodes()) > 1
+    ), "This test requires at least 2 GPUs"
+
+    actor_cls = TorchTensorWorker.options(num_cpus=0, num_gpus=1)
+
+    num_workers = 2
+    workers = [actor_cls.remote() for _ in range(num_workers)]
+
+    shape = (10,)
+    dtype = torch.float16
+
+    with InputNode() as inp:
+        computes = [
+            worker.compute_with_tuple_args.bind(inp, i)
+            for i, worker in enumerate(workers)
+        ]
+        collectives = collective.allreduce.bind(computes, ReduceOp.SUM)
+        dag = workers[0].recv.bind(collectives[0])
+
+    compiled_dag = dag.experimental_compile()
+
+    for i in range(3):
+        ref = compiled_dag.execute(
+            [(shape, dtype, idx + 1 + i) for idx in range(num_workers)]
+        )
+        result = ray.get(ref)
+        reduced_val = (num_workers + 1) * num_workers / 2 + i * num_workers
+        assert result == (reduced_val, shape, dtype)
+
+    compiled_dag.teardown()
+
+
+@pytest.mark.parametrize("ray_start_regular", [{"num_cpus": 4}], indirect=True)
+def test_torch_tensor_nccl_all_reduce_custom_comm(ray_start_regular):
+    """
+    Test all-reduce works with a custom communicator.
+    """
+    if not USE_GPU:
+        pytest.skip("NCCL tests require GPUs")
```

Comment:
It would be great if you can use the custom communicator so that these tests do not require GPU.

Reply:
I tried changing it to a non-GPU communicator, but there is a check in `_init_nccl_group` (`torch_tensor_nccl_channel.py:526`) that requires every actor to have a GPU. There is also an assertion in `_do_init_nccl_group` (`torch_tensor_nccl_channel.py:443`) that checks the same thing.

Reply:
You can start Ray with mock GPU types just by setting `"num_gpus": 4` in the pytest fixture instead of `"num_cpus": 4`.